src/arrow1/reader_async.rs

```diff
@@ -1,15 +1,569 @@
+use futures::channel::oneshot;
+use futures::future::BoxFuture;
+use parquet::arrow::ProjectionMask;
+use parquet::schema::types::SchemaDescriptor;
+use std::ops::Range;
 use std::sync::Arc;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::spawn_local;
 
-use crate::arrow1::error::Result;
-use crate::common::fetch::{create_reader, get_content_length};
+use crate::arrow1::error::{Result, WasmResult, ParquetWasmError};
+use crate::common::fetch::{
+    create_reader, get_content_length, range_from_end, range_from_start_and_length,
+};
 
 use arrow::ipc::writer::StreamWriter;
-use futures::StreamExt;
-use parquet::arrow::async_reader::{ParquetRecordBatchStream, ParquetRecordBatchStreamBuilder};
+use arrow_wasm::arrow1::{RecordBatch, Table};
+use bytes::Bytes;
+use futures::TryStreamExt;
+use futures::{stream, FutureExt, StreamExt};
+use parquet::arrow::arrow_reader::ArrowReaderMetadata;
+use parquet::arrow::async_reader::{
+    AsyncFileReader, ParquetRecordBatchStream, ParquetRecordBatchStreamBuilder,
+};
 
 use async_compat::{Compat, CompatExt};
-use parquet::file::metadata::FileMetaData;
+use parquet::file::footer::{decode_footer, decode_metadata};
+use parquet::file::metadata::{FileMetaData, ParquetMetaData};
 use range_reader::RangedAsyncReader;
+use reqwest::Client;
+use wasm_bindgen::prelude::*;
+use async_trait::async_trait;
+
+
+#[async_trait]
+trait SharedIO<T: AsyncFileReader + Unpin + Sync + Clone + 'static> {
+    fn generate_builder(reader: &T, meta: &ArrowReaderMetadata, batch_size: &usize, projection_mask: &Option<ProjectionMask>) -> ParquetRecordBatchStreamBuilder<T>{
+        let builder = ParquetRecordBatchStreamBuilder::new_with_metadata(
+            reader.clone(),
+            meta.clone(),
+        )
+        .with_batch_size(batch_size.clone())
+        .with_projection(projection_mask.as_ref().unwrap_or(&ProjectionMask::all()).clone());
+        builder
```

Comment:
I am not eager to add expression parsing or a DSL. For geoparquet wasm (being developed in https://github.com/geoarrow/geoarrow-rs), I think row filtering based on a bounding box is appropriate, but otherwise for now I don't want to add arbitrary row filtering if I can help it, pushing that to the end user in JS